File: python/WatchingCpl.java
import java.util.*;
class Codechef {

    // find sum
    static long sum(long a[]) {
        long sum=0;
        for (int i=0; i<a.length; i++)
            sum += a[i];
        return sum;
    }

    // check if the subset of a given sum exists
    static int dp(long a[], int i, int total) {
        Hashtable<String, Integer> mem = new Hashtable<>();
        return checkSubset(a, i, total, mem);
    }

    static int checkSubset(long a[], int i, int total, Hashtable<String, Integer> mem) {
        String key = total + ":" + i;
        if (mem.get(key) != null)
            return mem.get(key);
        if (total == 0)
            return 1;
        else if (total < 0 || i < 0)
            return 0;
        int value;
        if (total < a[i])
            value = checkSubset(a, i-1, total, mem);
        else 
            value = checkSubset(a, i-1, total, mem) | checkSubset(a, i-1, (int)(total-a[i]), mem);
        mem.put(key, value);
        return value;
    }
    public static void main(String[] args) {
        try(Scanner sc = new Scanner (System.in)) {
            int t = sc.nextInt();
            while(t-- > 0) {
                int n = sc.nextInt();
                int k = sc.nextInt();
                long h[] = new long[n];
                for (int i=0; i<n; i++) h[i] = sc.nextLong();

                // logic begins here
                if (n == 1) System.out.println(-1);
                else {
                    // sort the array
                    Arrays.sort(h);
                    long a[] = new long[n];
                    for (int i=0; i<n; i++)
                        a[i] = h[n-i-1];
                    
                    // max >= k
                    if (a[0] >= k) {
                        boolean flag = false;
                        long s = 0;
                        for (int i=1; i<n; i++) {
                            s += a[i];
                            if (s >= k) {
                               flag = true;
                               System.out.println(i+1);
                               break;
                            }
                        }
                        if (flag) System.out.println(-1);
                    }else {
                        if (sum(a) == (k << 1)) { // sum = 2k
                            if (dp(a, k, a.length-1) == 1)
                                System.out.println(n);
                            else
                                System.out.println(-1);
                        }else {
                            long s = 0;
                            ArrayList<Long> l = new ArrayList<>();
                            for (int i=0; i<n; i++) {
                                s += a[i];
                                if (s >= 2*k) {
                                    for (int j=0; j<=i; j++) 
                                        l.add(a[j]);
                                    break;
                                }
                            }
                            long cl[] = new long[l.size()];
                            for (int i=0; i<l.size(); i++)
                                cl[i] = l.get(i);
                            
                            if (dp(cl, k, cl.length-1) == 1)
                                System.out.println(cl.length);
                            else {
                                
                            }
                        }
                    }
                }
            }
        }
    }
}

File: python/watchingCPL.py
def checkSubset(h, total, i, mem):
    key = str(total) +":"+str(i)
    if key in mem:
        return mem[key]
    if total == 0:
        return True
    elif i < 0 or total < 0:
        return False
    elif total < h[i]:
        value =  checkSubset(h, total, i-1, mem)
    else:
        value =  checkSubset(h, total-h[i], i-1, mem) or checkSubset(h, total, i-1, mem)
    mem[key] = value
    return mem[key]

def dp(h, total, i):
    mem = {}
    return checkSubset(h, total, i, mem)

if __name__ == '__main__':
    t = int(input(''))
    while t > 0:
        s = list(map(int, input('').split(' ')))
        n = s[0]
        k = s[1]
        lt = []
        s = list(map(int, input('').split(' ')))
        for inp in s:
            lt.append(inp)
        if n == 1:
            print(-1)
        elif sum(lt) < (k << 1):  #sum < 2*k
            print(-1)
        else:
            lt = sorted(lt, reverse=True)
            if (lt[0] >= k):   #list[max] >= k
                flag = 0
                s = 0
                for i in range(1, n):
                    s += lt[i]
                    if (s >= k):
                        print(i+1)
                        flag = 1
                        break
                if flag == 0:
                    print(-1)
            else:
                if sum(lt) == (k << 1):
                    # i=n-1
                    # flag=False
                    # search = k-lt[0]
                    # while search > 0 and i >= 0:
                    #     if (lt.count(search) >= 1):
                    #         flag = True
                    #         break
                    #     else:
                    #         search = search - lt[i]
                    #         i -= 1
                    # if flag:
                    #     print(n)
                    # else:
                    #     print(-1)
                    if dp(lt, k, len(lt)-1):
                        print(n)
                    else:
                        print(-1)
                else:
                    s = 0
                    l = []
                    for i in range(n):
                        s += lt[i]
                        if (s >= 2*k):
                            for j in range(i+1):
                                l.append(lt[j])
                            break
                    # checking for sum == 2k
                    # if (sum(l) == (k << 1)):
                    #     if dp(l, k, len(l)-1):
                    #         print(len(l))
                    # else:
                    if dp(l, k, len(l)-1):
                            print(len(l))
                    else:
                        # print("l:", l)
                        l1 = [l[0]]
                        s = l[0]
                        idx = -1
                        for i in range(len(l)-1, 0, -1):
                            s += l[i]
                            l1.append(l[i])
                            if (s >= k):
                                idx = i
                                break
                        # print("l1:", l1)
                        # print("idx:", idx)
                        l2 = []
                        for i in range(1, idx):
                            l2.append(l[i])
                        # print("l2:", l2)
                        if (sum(l2) >= k):
                            print(len(l2) + len(l1))
                        else:
                            flag = False
                            s = sum(l2)
                            for i in range(len(l)-1, n):
                                s += lt[i]
                                l2.append(lt[i])
                                if s >= k:
                                    flag = True
                                    break
                            if flag:
                                print(len(l2) + len(l1))
                            else:
                                print(-1)
        t -= 1